python/ray/data/_internal/execution/resource_manager.py

```diff
@@ -121,3 +152,135 @@ def get_downstream_fraction(self, op: PhysicalOperator) -> float:
+        # Resource limits for each operator.
+        self._op_limits: Dict[PhysicalOperator, ExecutionResources] = {}
+        self._cached_global_limits = ExecutionResources.zero()
+
+    def _on_global_limits_updated(self, global_limits: ExecutionResources):
+        if len(self._eligible_ops) == 0:
+            return
+
+        self._total_shared = global_limits.scale(1.0 - self._reservation_ratio)
+
+        default_reserved = global_limits.scale(
+            self._reservation_ratio / len(self._eligible_ops)
+        )
+        for op in self._eligible_ops:
+            # Make sure the reserved resources are at least to allow
+            # one task.
+            self._op_reserved[op] = default_reserved.max(
+                op.incremental_resource_usage()
+            )
+
+    def update_usages(self):
+        if len(self._eligible_ops) == 0:
+            return
+
+        global_limits = self._resource_manager.get_global_limits()
+        if global_limits != self._cached_global_limits:
+            self._on_global_limits_updated(global_limits)
+            self._cached_global_limits = global_limits
+
+        self._op_limits.clear()
+        # Remaining of shared resources.
+        remaining_shared = self._total_shared
+        for op in self._resource_manager._topology:
+            op_usage = self._resource_manager.get_op_usage(op)
+            if op in self._eligible_ops:
+                op_reserved = self._op_reserved[op]
+                # How much of the reserved resources are remaining.
+                op_reserved_remaining = op_reserved.subtract(op_usage).max(
+                    ExecutionResources.zero()
+                )
+                self._op_limits[op] = op_reserved_remaining
+                # How much of the reserved resources are exceeded.
+                # If exceeded, we need to subtract from the remaining shared resources.
+                op_reserved_exceeded = op_usage.subtract(op_reserved).max(
+                    ExecutionResources.zero()
+                )
+                remaining_shared = remaining_shared.subtract(op_reserved_exceeded)
+            else:
+                # For non-eligible ops, we still need to subtract
+                # their usage from the remaining shared resources.
+                remaining_shared = remaining_shared.subtract(op_usage)
+
+        shared_divided = remaining_shared.max(ExecutionResources.zero()).scale(
+            1.0 / len(self._eligible_ops)
+        )
+        for op in self._eligible_ops:
+            self._op_limits[op] = self._op_limits[op].add(shared_divided)
+            # We don't limit GPU resources, as not all operators
+            # use GPU resources.
+            self._op_limits[op].gpu = float("inf")
+
+    def get_op_limits(self, op: PhysicalOperator) -> ExecutionResources:
+        if op in self._op_limits:
+            return self._op_limits[op]
+        else:
+            return ExecutionResources.inf()
```

Comment:
Nit: Optionally, 
```suggestion
        return self._op_limits.get(op, ExecutionResources.inf())
```

Reply:
Not a big deal, but the current way will be slightly more efficient, as `inf()` now still returns a new object each time.